File: momors/crates/momors-core/src/bracket.rs
//! 括弧類の構造分類テーブル。
//!
//! 括弧まわりの分かち書きを当てるには、括弧を素通しの記号として他の記号と
//! 同列にモデルへ流すだけでは足りない。個々の括弧字形（「『""''【 など）は
//! bigram/trigram特徴量にとって希少すぎて汎化しないため、特徴量計算だけは
//! Role×Treatmentを3値に圧縮したトークンに置き換えたビューから取る（実処理は
//! [`crate::prediction`] にある）。出力・ctype分岐・境界判定は常に実文字の
//! まま ─ 読みは他の記号と同じくbypass、直後のマスあけは境界モデルに委ねる
//! （`CharType::skips_boundary_check` 参照）。
//!
//! # 中身を本文の流れに残してよいか（[`Treatment`]）
//!
//! 括弧種によって扱いが2通りに分かれる。判定基準は **中身が文の流れの一部か** :
//!
//! - 引用 `彼は『走れメロス』を読む`: 中身は文の項なので流れの中にいる。
//!   括弧ごとそのまま本文に残して推論しても自然文のままで、モデルへの問い
//!   （`朝|おはよう` を空けるか）も正しい問いになる → [`Inline`]。
//! - 注釈 `オケ（オーケストラ）の団員`: 中身は傍注で流れの外にいる。その場に
//!   残すと `オケオーケストラ` のような実在しない隣接ができ、特徴量ウィンドウ
//!   が偽の文脈をまたぐ。span ごと本文から抜いて独立に推論すれば、本文は
//!   `オケの団員` と自然なままで、偽の隣接が最初から生じない → [`Aside`]。
//!
//! 分類は少数なので TOML ではなく本モジュールに直接持つ（core を toml 非依存に
//! 保つ）。点字変換テーブル（momors-braille）側の `class` とは関心が別
//! （あちらはセル隣接のスペース、こちらは本文の再構成）であり、モジュール
//! 独立性を優先して意図的に別管理にしている。
//!
//! [`Inline`]: Treatment::Inline
//! [`Aside`]: Treatment::Aside

/// 括弧の食いつき方向。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Role {
    /// 開き括弧。直後の文字に密着する（「 → 「オ）。
    Open,
    /// 閉じ括弧。直前の文字に密着する（」 → ヨー」）。
    Close,
}

/// 括弧の中身を本文と一緒に推論してよいか。モジュール冒頭の説明を参照。
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Treatment {
    /// 中身は文の流れの一部（引用系）。括弧文字も本文から外さず、実文字の
    /// まま推論に乗せる（特徴量計算だけトークン置換する）。
    Inline,
    /// 中身は流れの外の傍注（注釈系）。span ごと本文から抜いて独立に推論し、
    /// 前の語へ食いつけて合成する。
    Aside,
}

/// `c` が括弧なら (役割, 中身の扱い) を返す。該当しなければ `None`。
///
/// 分類が実運用で合わないと分かれば、ここの割り当てを直すだけで調整できる。
pub(crate) fn lookup(c: char) -> Option<(Role, Treatment)> {
    use Role::*;
    use Treatment::*;
    Some(match c {
        // 引用系: 中身は文の項
        '「' => (Open, Inline),
        '」' => (Close, Inline),
        '『' => (Open, Inline),
        '』' => (Close, Inline),
        '“' => (Open, Inline),
        '”' => (Close, Inline),
        '‘' => (Open, Inline),
        '’' => (Close, Inline),
        '【' => (Open, Inline),
        '】' => (Close, Inline),
        // 注釈・挿入系: 中身は傍注
        '（' => (Open, Aside),
        '）' => (Close, Aside),
        '(' => (Open, Aside),
        ')' => (Close, Aside),
        '｛' => (Open, Aside),
        '｝' => (Close, Aside),
        _ => return None,
    })
}

/// `open` と `close` が同じ括弧ペアか。
pub(crate) fn is_pair(open: char, close: char) -> bool {
    matches!(
        (open, close),
        ('「', '」')
            | ('『', '』')
            | ('“', '”')
            | ('‘', '’')
            | ('（', '）')
            | ('(', ')')
            | ('｛', '｝')
            | ('【', '】')
    )
}

/// 特徴量計算専用の圧縮アイデンティティ・トークン（Private Use Area）。
///
/// 個々の括弧字形（「『""''【 など）はbigram/trigram特徴量にとって希少すぎて
/// 汎化しない。そこで特徴量計算専用の系列（出力・ctype分岐・境界判定を決める
/// `text` 側とは別のビュー）でだけ、括弧の位置をこの3値の圧縮トークンに
/// 置き換える。ctypeは既存のSymbolOpen/SymbolClose（get_char_type）のままで
/// よい（役割はそちらで十分圧縮済み）。
///
/// `INLINE_OPEN_TOKEN`/`INLINE_CLOSE_TOKEN` は `text` 側にも同じ位置に実文字
/// （括弧そのもの）が1:1で存在する ─ Inline括弧は本文から外さないため、その
/// 行自身が実際に読み/境界の対象になる（読みはbypass、境界は境界モデル）。
/// 一方 `ASIDE_TOKEN` は span 丸ごと1文字に圧縮したプレースホルダで、`text`
/// 側には対応する実文字が無い（Aside の中身は独立した副文として本文から
/// 完全に抜かれる）。
///
/// momo-py/src/momo_py/bracket.py と同じコードポイントを使うこと。
/// ずれるとモデルの語彙（char_s等の特徴量キー）がPython/Rustで食い違う。
pub(crate) const INLINE_OPEN_TOKEN: char = '\u{E000}';
pub(crate) const INLINE_CLOSE_TOKEN: char = '\u{E001}';
pub(crate) const ASIDE_TOKEN: char = '\u{E002}';

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn quotation_is_inline() {
        for c in "「」『』“”‘’【】".chars() {
            assert_eq!(lookup(c).map(|(_, t)| t), Some(Treatment::Inline), "{c}");
        }
    }

    #[test]
    fn annotation_is_aside() {
        for c in "（）()｛｝".chars() {
            assert_eq!(lookup(c).map(|(_, t)| t), Some(Treatment::Aside), "{c}");
        }
    }

    #[test]
    fn roles() {
        for c in "「『“‘（(｛【".chars() {
            assert_eq!(lookup(c).map(|(r, _)| r), Some(Role::Open), "{c}");
        }
        for c in "」』”’）)｝】".chars() {
            assert_eq!(lookup(c).map(|(r, _)| r), Some(Role::Close), "{c}");
        }
    }

    #[test]
    fn pairs_match() {
        assert!(is_pair('（', '）'));
        assert!(is_pair('「', '」'));
        assert!(!is_pair('（', '」'));
        assert!(!is_pair('「', '）'));
    }

    #[test]
    fn non_brackets_are_none() {
        assert_eq!(lookup('あ'), None);
        assert_eq!(lookup('。'), None);
        assert_eq!(lookup('、'), None);
        assert_eq!(lookup('…'), None);
        assert_eq!(lookup('A'), None);
    }

    #[test]
    fn identity_tokens_match_python_bracket_py() {
        // momo-py/src/momo_py/bracket.py の INLINE_OPEN_TOKEN/INLINE_CLOSE_TOKEN/
        // ASIDE_TOKEN と同じコードポイントであること（ずれるとモデルの語彙が
        // Python/Rustで食い違う）。
        assert_eq!(INLINE_OPEN_TOKEN as u32, 0xE000);
        assert_eq!(INLINE_CLOSE_TOKEN as u32, 0xE001);
        assert_eq!(ASIDE_TOKEN as u32, 0xE002);
    }
}

File: momo-py/src/momo_py/bracket.py
"""括弧類の構造分類テーブル（momors-core/src/bracket.rs の移植）。

Rust側(momors-core)の推論時strip/splice処理と同じ分類をPython学習側でも
使えるようにするための、依存の少ない小さなモジュール。判定基準・分類は
bracket.rsと同一に保つこと（どちらかを直したら両方直す）。

- Inline（引用系「」『』""''【】）: 中身は文の項。中身は本文の流れに残す。
- Aside（注釈系（）()｛｝）: 中身は傍注。span ごと本文から抜く。
"""

from enum import Enum
from typing import List, Optional, Tuple

from .features import MORA_SPLIT


class Role(str, Enum):
    """括弧の食いつき方向。"""

    OPEN = "OPEN"
    CLOSE = "CLOSE"

    def __str__(self) -> str:
        return self.value


class Treatment(str, Enum):
    """括弧の中身を本文と一緒に扱ってよいか。"""

    INLINE = "INLINE"
    ASIDE = "ASIDE"

    def __str__(self) -> str:
        return self.value


# bracket.rs の lookup() と同一の対応表。
_TABLE = {
    # 引用系: 中身は文の項
    "「": (Role.OPEN, Treatment.INLINE),
    "」": (Role.CLOSE, Treatment.INLINE),
    "『": (Role.OPEN, Treatment.INLINE),
    "』": (Role.CLOSE, Treatment.INLINE),
    "“": (Role.OPEN, Treatment.INLINE),  # “
    "”": (Role.CLOSE, Treatment.INLINE),  # ”
    "‘": (Role.OPEN, Treatment.INLINE),  # ‘
    "’": (Role.CLOSE, Treatment.INLINE),  # ’
    "【": (Role.OPEN, Treatment.INLINE),
    "】": (Role.CLOSE, Treatment.INLINE),
    # 注釈・挿入系: 中身は傍注
    "（": (Role.OPEN, Treatment.ASIDE),
    "）": (Role.CLOSE, Treatment.ASIDE),
    "(": (Role.OPEN, Treatment.ASIDE),
    ")": (Role.CLOSE, Treatment.ASIDE),
    "｛": (Role.OPEN, Treatment.ASIDE),
    "｝": (Role.CLOSE, Treatment.ASIDE),
}

_PAIRS = {
    ("「", "」"),
    ("『", "』"),
    ("“", "”"),
    ("‘", "’"),
    ("（", "）"),
    ("(", ")"),
    ("｛", "｝"),
    ("【", "】"),
}


def lookup(c: str) -> Optional[Tuple[Role, Treatment]]:
    """`c` が括弧なら (役割, 中身の扱い) を返す。該当しなければ None。"""
    return _TABLE.get(c)


# 対応表に載っている全括弧文字（文中に括弧が含まれるかの判定に使う）。
ALL_BRACKET_CHARS = frozenset(_TABLE.keys())


# 特徴量計算専用の圧縮アイデンティティ・トークン（Private Use Area）。
#
# 個々の括弧字形（「『""''【 など）はbigram/trigram特徴量にとって希少すぎて
# 汎化しない。かといって完全にstripすると、隣接文字のkanji_run/kanji_pos_first/
# bigram/trigramが「括弧がそこにあった」ことを一切知りえず、区切りとして
# 認識できない（実測: momo-inspect trace で「、」が読み分けに効く主因は
# bi_p1_s/type_tri_p2_p1_s のような隣接ベースの特徴量だった）。
#
# そこで、特徴量計算専用の系列（出力・マスあけを決めるstrip/Atom/splice経路とは
# 別）でだけ、括弧の位置をこの3値の圧縮トークンに置き換えて残す。ctypeは
# 既存のSYMBOL_OPEN/SYMBOL_CLOSE（get_char_type）のままでよい（役割はそちらで
# 十分圧縮済み）。この位置自体には読み/境界のラベルは立てない。
#
# 将来Rust側（momors-core/src/bracket.rs）でも同じコードポイントを使うこと。
# ずれるとモデルの語彙（char_s等の特徴量キー）がPython/Rustで食い違う。
INLINE_OPEN_TOKEN = ""
INLINE_CLOSE_TOKEN = ""
ASIDE_TOKEN = ""


def is_pair(open_c: str, close_c: str) -> bool:
    """`open_c` と `close_c` が同じ括弧ペアか。"""
    return (open_c, close_c) in _PAIRS


# 学習データ用の行表現: [char, label, ctype, orig_idx, name_flag, ...]
Row = List[str]
Sentence = List[Row]


def _has_space(label: str) -> bool:
    return MORA_SPLIT in label


def _set_space(label: str, want_space: bool) -> str:
    base = label.replace(MORA_SPLIT, "")
    return base + MORA_SPLIT if want_space else base


def _with_label(row: Row, new_label: str) -> Row:
    return row[:1] + [new_label] + row[2:]


def _with_char_and_label(row: Row, new_char: str, new_label: str) -> Row:
    return [new_char, new_label] + row[2:]


# trainer.py側でこのラベルの行を実際の学習例（Y_read/Y_boundary）から除外する
# ためのセンチネル。source_seq（ひいてはbigram/trigram/kanji_run等の特徴量
# 計算）には残す＝隣接する実文字が「括弧がここにあった」ことを認識できる
# ようにするが、この行自体には読み/境界を予測させない。
#
# ASIDE括弧（開き括弧の行）にのみ使う。中身がspanごと本文から抜かれて独立文
# になるため、開き括弧の行自体は「そこにASIDEがあった」ことを示す特徴量専用の
# プレースホルダでしかなく、実際の出力・マスあけの決定対象になり得ない
# （Rust推論側のextract_aside_spans/splice_atomsが別途決める）。
#
# INLINE括弧には使わない（tokenize_for_training参照）。中身と一緒に本文の
# 流れに残るため、開き・閉じの行自体もRust推論側で実際の境界判定対象になる
# ─ その行自身のラベル（+Sの有無）をそのまま学習に使う。
LABEL_CONTEXT_ONLY = "@"


def strip_for_training(rows: Sentence) -> Tuple[Sentence, List[Sentence]]:
    """学習データの行列から、bracket.rs（momors-core）の推論時strip/splice処理と
    同じ規則で括弧を取り除く。

    従来のtrainer.pyは括弧を生のまま学習データに残していたため、Rust推論側が
    常に括弧をstripしてから予測する実際の入力と食い違っていた
    （bigram/trigram文脈やAsideの独立推論構造が学習/推論で不一致）。
    utilities/bracket_boundary_experiment.py での実データ検証で、この不整合の
    解消だけで実際のBOUNDARY（分かち書き）誤り2件が直ることを確認済み。

    - INLINE括弧（「」『』""''【】）: その場でstrip。閉じ括弧行に付いていた
      +S（空白ラベル）は、strip後に地の文で直前に来る実文字の行へ移す
      （bracket.rsのstrip_bracket_charsは括弧を跨いでモデルの空白判断に触れない
      ため、学習側も同じ隣接関係で+Sを持たせる必要がある）。
    - ASIDE括弧（（）()｛｝）: spanごと本文から抜き取り、中身を独立した副文
      として返す（Rust側predict_with_bracketsの再帰独立推論に対応）。閉じ括弧
      行の+Sは、開き括弧の直前に来る実文字の行へ移す。

    戻り値: (stripしたメイン文, 副文のリスト。入れ子Asideは再帰的に平坦化)
    """
    out_rows: Sentence = []
    sub_sentences: List[Sentence] = []
    i, n = 0, len(rows)
    while i < n:
        row = rows[i]
        ch = row[0]
        info = lookup(ch) if len(ch) == 1 else None

        if info is None:
            out_rows.append(row)
            i += 1
            continue

        role, treatment = info

        if treatment == Treatment.INLINE:
            if role == Role.OPEN:
                if _has_space(row[1]):
                    print(f"⚠️  開き括弧行に+S(未対応パターン、無視): {row}")
            else:  # CLOSE
                if out_rows and _has_space(row[1]):
                    out_rows[-1] = _with_label(out_rows[-1], _set_space(out_rows[-1][1], True))
                elif _has_space(row[1]):
                    print(f"⚠️  文頭付近の閉じ括弧+Sを移設できません(無視): {row}")
            i += 1
            continue

        # Aside: 対応する閉じまでdepth追跡して span ごと抜き取る。
        depth, j = 1, i + 1
        content_rows: Sentence = []
        close_row: Optional[Row] = None
        while j < n:
            ch2 = rows[j][0]
            info2 = lookup(ch2) if len(ch2) == 1 else None
            if info2 is not None and info2[1] == Treatment.ASIDE:
                if info2[0] == Role.OPEN:
                    depth += 1
                    content_rows.append(rows[j])
                else:
                    depth -= 1
                    if depth == 0:
                        close_row = rows[j]
                        j += 1
                        break
                    content_rows.append(rows[j])
            else:
                content_rows.append(rows[j])
            j += 1

        if close_row is None:
            print(f"⚠️  対応する閉じ括弧が見つかりません(素通し): {row}")
            out_rows.append(row)
            i += 1
            continue

        if out_rows and _has_space(close_row[1]):
            out_rows[-1] = _with_label(out_rows[-1], _set_space(out_rows[-1][1], True))
        elif _has_space(close_row[1]):
            print(f"⚠️  文頭の注釈括弧の閉じ+Sを移設できません(無視): {close_row}")

        if content_rows:
            sub_main, sub_subs = strip_for_training(content_rows)
            sub_sentences.append(sub_main)
            sub_sentences.extend(sub_subs)

        i = j

    return out_rows, sub_sentences


def tokenize_for_training(rows: Sentence) -> Tuple[Sentence, List[Sentence]]:
    """学習データの行列に対し、括弧の行を削除せず残しつつ、charを特徴量計算
    専用の圧縮トークン（INLINE_OPEN_TOKEN/INLINE_CLOSE_TOKEN/ASIDE_TOKEN）に
    置き換える（strip_for_trainingの後継、圧縮アイデンティティ・トークン方式）。

    strip_for_trainingとの違い: 括弧の行そのものは削除しない。char列だけを
    トークンへ差し替える。個々の括弧字形はbigram/trigramにとって希少すぎて
    汎化しないが、Role×Treatmentを3値に圧縮したトークンなら汎化する
    （実測: momo-inspect traceで「、」が読み分けに効く主因はbi_p1_s/
    type_tri_p2_p1_sのような隣接ベースの特徴量だった。kanji_run/
    kanji_pos_firstはsource_seq上の隣接をそのまま辿る実装のため、
    strip_for_trainingのように行ごと削除すると括弧の存在を一切認識できない）。
    utilities/bracket_identity_token_experiment.pyでの実データ検証で、
    READING誤り2件（海→ウミ、楽→ガッ）が直ることを確認済み。

    - INLINE括弧: その場でトークンに置換するだけで、ラベル（`+S`の有無を
      含む）には一切触れない。中身と一緒に本文の流れに残り、Rust推論側
      （momors-core::prediction）でも実文字のまま境界判定の対象になる
      （strip/spliceでは外さない）ため、開き・閉じの行自身が持つ`+S`を
      そのまま学習に使えばよく、直前行への移設は不要（旧strip_for_training
      時代の名残）。
    - ASIDE括弧: 開き括弧の行だけを`ASIDE_TOKEN`＋`LABEL_CONTEXT_ONLY`に
      置換し、閉じ括弧の行と中身は本文から除く。中身は独立した副文として
      返す（Rust側predict_with_bracketsの再帰独立推論に対応）。中身は
      span ごと本文から抜かれるため、閉じ括弧の`+S`はRust側と同じく
      開き括弧の直前に来る実文字行へ移す（Aside側は従来通り）。

    戻り値: (トークン化したメイン文, 副文のリスト。入れ子Asideは再帰的に平坦化)
    """
    out_rows: Sentence = []
    sub_sentences: List[Sentence] = []
    i, n = 0, len(rows)
    while i < n:
        row = rows[i]
        ch = row[0]
        info = lookup(ch) if len(ch) == 1 else None

        if info is None:
            out_rows.append(row)
            i += 1
            continue

        role, treatment = info

        if treatment == Treatment.INLINE:
            token = INLINE_OPEN_TOKEN if role == Role.OPEN else INLINE_CLOSE_TOKEN
            out_rows.append(_with_char_and_label(row, token, row[1]))
            i += 1
            continue

        # Aside: 対応する閉じまでdepth追跡。開き行をASIDEトークンに置換し、
        # 閉じ行・中身は本文から除く（中身は独立した副文として再帰処理）。
        depth, j = 1, i + 1
        content_rows: Sentence = []
        close_row: Optional[Row] = None
        while j < n:
            ch2 = rows[j][0]
            info2 = lookup(ch2) if len(ch2) == 1 else None
            if info2 is not None and info2[1] == Treatment.ASIDE:
                if info2[0] == Role.OPEN:
                    depth += 1
                    content_rows.append(rows[j])
                else:
                    depth -= 1
                    if depth == 0:
                        close_row = rows[j]
                        j += 1
                        break
                    content_rows.append(rows[j])
            else:
                content_rows.append(rows[j])
            j += 1

        if close_row is None:
            print(f"⚠️  対応する閉じ括弧が見つかりません(素通し): {row}")
            out_rows.append(row)
            i += 1
            continue

        if out_rows and _has_space(close_row[1]):
            out_rows[-1] = _with_label(out_rows[-1], _set_space(out_rows[-1][1], True))
        elif _has_space(close_row[1]):
            print(f"⚠️  文頭の注釈括弧の閉じ+Sを移設できません(無視): {close_row}")

        out_rows.append(_with_char_and_label(row, ASIDE_TOKEN, LABEL_CONTEXT_ONLY))

        if content_rows:
            sub_main, sub_subs = tokenize_for_training(content_rows)
            sub_sentences.append(sub_main)
            sub_sentences.extend(sub_subs)

        i = j

    return out_rows, sub_sentences
